pkg/controller/mysql/grant/reconciler.go

```diff
@@ -0,0 +1,324 @@
+)
+
+var (
+	grantRegex = regexp.MustCompile("^GRANT (.+) ON `(.+)`\\.\\* TO .+")
+)
+
+// Setup adds a controller that reconciles Grant managed resources.
+func Setup(mgr ctrl.Manager, l logging.Logger) error {
+	name := managed.ControllerName(v1alpha1.GrantGroupKind)
+
+	t := resource.NewProviderConfigUsageTracker(mgr.GetClient(), &v1alpha1.ProviderConfigUsage{})
+	r := managed.NewReconciler(mgr,
+		resource.ManagedKind(v1alpha1.GrantGroupVersionKind),
+		managed.WithExternalConnecter(&connector{kube: mgr.GetClient(), usage: t, newDB: mysql.New}),
+		managed.WithInitializers(managed.NewDefaultProviderConfig(mgr.GetClient()), managed.NewNameAsExternalName(mgr.GetClient())),
+		managed.WithReferenceResolver(managed.NewAPISimpleReferenceResolver(mgr.GetClient())),
+		managed.WithLogger(l.WithValues("controller", name)),
+		managed.WithRecorder(event.NewAPIRecorder(mgr.GetEventRecorderFor(name))))
+
+	return ctrl.NewControllerManagedBy(mgr).
+		Named(name).
+		For(&v1alpha1.Grant{}).
+		Complete(r)
+}
+
+type connector struct {
+	kube  client.Client
+	usage resource.Tracker
+	newDB func(creds map[string][]byte) xsql.DB
+}
+
+func (c *connector) Connect(ctx context.Context, mg resource.Managed) (managed.ExternalClient, error) {
+	cr, ok := mg.(*v1alpha1.Grant)
+	if !ok {
+		return nil, errors.New(errNotGrant)
+	}
+
+	if err := c.usage.Track(ctx, mg); err != nil {
+		return nil, errors.Wrap(err, errTrackPCUsage)
+	}
+
+	// ProviderConfigReference could theoretically be nil, but in practice the
+	// DefaultProviderConfig initializer will set it before we get here.
+	pc := &v1alpha1.ProviderConfig{}
+	if err := c.kube.Get(ctx, types.NamespacedName{Name: cr.GetProviderConfigReference().Name}, pc); err != nil {
+		return nil, errors.Wrap(err, errGetPC)
+	}
+
+	// We don't need to check the credentials source because we currently only
+	// support one source (MySQLConnectionSecret), which is required and
+	// enforced by the ProviderConfig schema.
+	ref := pc.Spec.Credentials.ConnectionSecretRef
+	if ref == nil {
+		return nil, errors.New(errNoSecretRef)
+	}
+
+	s := &corev1.Secret{}
+	if err := c.kube.Get(ctx, types.NamespacedName{Namespace: ref.Namespace, Name: ref.Name}, s); err != nil {
+		return nil, errors.Wrap(err, errGetSecret)
+	}
+
+	return &external{
+		db:   c.newDB(s.Data),
+		kube: c.kube,
+	}, nil
+}
+
+type external struct {
+	db   xsql.DB
+	kube client.Client
+}
+
+func (c *external) Observe(ctx context.Context, mg resource.Managed) (managed.ExternalObservation, error) {
+	cr, ok := mg.(*v1alpha1.Grant)
+	if !ok {
+		return managed.ExternalObservation{}, errors.New(errNotGrant)
+	}
+
+	username := *cr.Spec.ForProvider.User
+	dbname := *cr.Spec.ForProvider.Database
+
+	privileges, result, err := c.getPrivileges(ctx, username, dbname)
+	if err != nil {
+		return managed.ExternalObservation{}, err
+	}
+	if result != nil {
+		return *result, nil
+	}
+
+	if !comparePrivileges(cr.Spec.ForProvider.Privileges, privileges) {
```

Comment:
Naming nit: Maybe something like `!equal` or `!privilegesEqual` rather than `!comparePrivileges` would more clearly indicate that this returns true if the privileges are equal.